platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,493 @@
+     * <p>
+     * This value SHALL be the hash of the active roster used for the round.
+     */
+    byte active_roster_hash = 2;
+}
+```
+
+It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
+This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, and an
+optional current Candidate Roster. Insertion of rosters will be controlled by ensuring that the acceptance of a new
+Candidate Roster invalidates and removes the current candidate roster. Therefore, the map is expected to have a maximum
+size of three elements.
+The new `RosterState` will store the `candidateRosterHash` and a list `roundRosterPairs`
+of pairs of round numbers and the active roster that was used for that round. The `roundRosterPairs` list will be
+updated only when the active roster changes.
+In the vast majority of cases, the list will contain only one element, the current active roster.
+Immediately following the adoption of a candidate roster, this list will have two elements — the previous active roster
+and the current active roster (which was the candidate roster).
+The choice of a list is suitable here as it is much cheaper to iterate over 2-3 elements comparing their `long` round
+numbers than it is to do a hash lookup.
+When we reach full Dynamic Address Book, this design choice may need to be revisited.
+
+If a `candidateRosterHash` hash entry already exist (hash collisions) in the map of Rosters, it will be discarded. That
+is, setting a candidate roster is an idempotent operation.
+
+One benefit of this indirection (using a Map instead of a Singleton) is that it avoids moving data around in the merkle
+state.
+Another benefit is that adoption trigger becomes straightforward (App sets the roster) with delineated
+responsibilities between the App and the platform.
+
+### Roster Validity
+
+A Roster is considered valid if it satisfies the following conditions:
+
+1. The roster must have at least one RosterEntry.
+2. At least one RosterEntry/ies must have a non-zero weight.
+3. All RosterEntry/ies must have a valid X509Certificate.
+4. All RosterEntry/ies must have a valid PairingPublicKey.
+5. All RosterEntry/ies must have at least one gossip Endpoint.
+6. The RosterEntry/ies must be specified in order of ascending node id.
+7. All ServiceEndpoint/s must have a valid IP address or domain name (mutually exclusive), and port.
+8. The roster must have a unique NodeId for each RosterEntry, and the NodeId key must match the NodeId value in the
+   corresponding RosterEntry.
+
+On the submission of a new Candidate Roster, the platform will validate the roster against these conditions.
+Note that a constructed `Roster` can be valid, but not accepted by the platform.
+For example, if a new candidate roster is set via the API, but its hash evaluates to the hash of the existing
+candidate roster, the new roster will be discarded. That is, the operation has no effect.
```

Comment:
Practically speaking, this just means the app submitted the same candidate roster twice?

Reply:
yes. We're only making it explicit that candidate roster submission will be an idempotent operation.